### DISCUSSION TITLE
How can I override CTRL+F keybinding?

### DISCUSSION THREAD

**Discussion by sontek** [PatrickF1/fzf.fish#108] [Archive] [answered]
CTRL+F is a standard key binding for most shells (sh, bash, zsh, fish) for moving forward a character.  So installing fzf has really screwed up my terminal usage since 20 years of memory means I need ctrl+f to do what its always done :)

I see in `.conf.d/fzf/` its declared  but I don't really want to edit that file since I'd like to be able to upgrade if things change in the future.   Is there an easy way to just say "reset ctrl+f to existing shell"?

I'm not great with fish yet, still learning the differences.   I saw in the docs you say to do:

```
set --universal fzf_fish_custom_keybindings
```

but `Do not try to set fzf_fish_custom_keybindings in your config.fish`  but I don't understand where I would set it if not in config.fish.  Then I don't understand how I can just remove your binding thats overriding the defaults.  I don't want to change the binding to something else, just remove it

**Comment by sontek** [comment]
```
bind --erase --all \cf
```

seems to be all you need for this.   I'll leave this ticket open for discussion since I think its bad practice to override such an important default keybinding as movement in shells but if you disagree we can close this.

**Comment by PatrickF1** [comment]
~Hi @sontek! Did you try https://github.com/PatrickF1/fzf.fish#using-custom-key-bindings?~

Oops didn't see your edit.

> I'm not great with fish yet, still learning the differences. I saw in the docs you say to do:
set --universal fzf_fish_custom_keybindings
but Do not try to set fzf_fish_custom_keybindings in your config.fish but I don't understand where I would set it if not in config.fish. Then I don't understand how I can just remove your binding thats overriding the defaults. I don't want to change the binding to something else, just remove it

Ah, you can include that command in your config.fish. What I mean is don't try to do `set fzf_fish_custom_keybindings` in your config.fish without the `--universal` flag. I'll clarify.

**Comment by sontek** [comment]
but that seems to overwrite all the default keybindings? Not just give the ability to customize?    My big concern is this is squashing an extremely important default keybinding for shells, so it doesn't seem to be a good default.   It removes the ability to move forward on the command line

**Comment by PatrickF1** [comment]
> but that seems to overwrite all the default keybindings? Not just give the ability to customize?

You're right. It's not ideal but I don't think it's a big deal because it's just 4 simple lines of code you can almost copy and paste. My hands are also tied by my previous (less informed) decisions. Now that probably thousands of people are using this plugin, I can't just change the defaults or settings on them.

> It removes the ability to move forward on the command line

Can you explain this statement more? How come you don't use the arrow keys? I'm trying to understand my users' setups and work styles so I can make key bindings better in the future

**Comment by sontek** [comment]
I think most people who are using the CLI are trying to stay on the homerow rather than using arrow keys, which is why ctrl+f and ctrl+b exist in the shells.   It is just an ergonomic thing so that my fingers don't have to move off and prevent me from typing. 

There are also other shortcuts that work the same way for better movement than arrow keys.  For example I use alt+f / alt+b as well for jumping words instead of characters.  Ctrl+A / Ctrl+E to go to front and back of a line.

**Comment by PatrickF1** [comment]
Ah you must be a pro Emacs user!

> I think most people who are using the CLI are trying to stay on the homerow rather than using arrow keys

I would disagree with this. I think most people use arrow keys, a small minority use vim-bindings, in which case they either use arrow keys or `h` and `l`, and a small minority use all of Emacs-bindings. So I think overall Ctrl+F isn't a HUGE problem, but it is a pretty significant problem that a handful of people like you have to deal with. I wish I had known earlier :(
Thanks for enlightening me. I'll think harder about a solution even if there's no clean one.

> Ctrl+A / Ctrl+E to go to front and back of a line

I do use these two as well as Ctrl+W and Alt+F. Just fo some reason, I never use Ctrl+F and Ctrl+B to move one letter at a time. In fact, most people I've pair programmed with, both in college, and at my jobs, don't.

**Comment by PatrickF1** [comment]
Btw @sontek, you mentioned you were new to Fish. I want to say this: welcome! I hope you'll love it and make it your new home. Since you're still learning Fish and the different variable types, may I recommend 
 this wonderful guide that got me started: [cookbook.fish](https://github.com/jorgebucaran/cookbook.fish)? Also, I recently made an edit to it just for you. See this section https://github.com/jorgebucaran/cookbook.fish#how-do-i-set-variables-in-fish

**Comment by angrypie** [comment]
Hey there. For me `Ctrl+F` is also confusing because it's `fish` default key binding to accept autosuggestion and I used to it.

**Comment by PatrickF1** [comment]
Try using the right arrow key for now. I’m thinking now of how I can fix
this without a rough transition for tons of people.

**Comment by PatrickF1** [comment]
@angrypie @sontek if I change it to Alt+F, would that work for you?

**Comment by sontek** [comment]
@PatrickF1 That would still stomp on default Fish key mappings as well as standard shell mappings.   

For fish specifically these are the two default bindings for movement:

```
bind --preset \ef forward-word
bind --preset \cf forward-char
```

Since these already had special meaning, during autocompletion (when there is no forward to go to) fish special cases them to do "forward total completion" and "forward next word in completion"

![Jan-24-2021 14-09-55](https://user-images.githubusercontent.com/151924/105639312-e147bd80-5e4d-11eb-8d2c-c4b16f48b9a3.gif)

I suspect that there are more fish (and standard shell users) who are used to using these standard keystrokes more than ones who have specialized their usage to this specific plugin.   The first thing I had to do when using this plugin was learn how to teach it not to step on those keybindings

I think what might be best is find some binding that isn't already in the presets and use that but give a hook to easily change it to whatever people want

**Comment by PatrickF1** [comment]
@sontek Thanks for the info! I was also unaware of that binding. Now I see it in the default bindings. What do you think about Ctrl+Alt+F then?

> I suspect that there are more fish (and standard shell users) who are used to using these standard keystrokes more than ones who have specialized their usage to this specific plugin. 

I can't be sure but I agree now, based on the volume of people asking me to change this. And for that reason I will change it.

**Comment by angrypie** [comment]
@sontek For me it's more convenient to be able just put `set --global ...` or something in my config to override few defaults.

**Comment by PatrickF1** [comment, marked as answer]
@sontek @angrypie The default binding is now Ctrl+Alt+F thanks to your feedback! Additionally, custom key bindings are now a breeze. See https://github.com/PatrickF1/fzf.fish/releases/tag/v7.0.
